internal/credentials/credentialls.go

```diff
@@ -0,0 +1,176 @@
+package credentials
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+	"math/big"
+
+	"golang.org/x/crypto/bcrypt"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	hivev1 "github.com/openshift/hive/apis/hive/v1"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	DefaultUser    = "kubeadmin"
+	kubeconfig     = "kubeconfig"
+	kubeadmincreds = "kubeadmincreds"
+)
+
+type Credentials struct {
+	client.Client
+	Log    logrus.FieldLogger
+	Scheme *runtime.Scheme
+}
+
+func (r *Credentials) EnsureKubeconfigSecret(ctx context.Context, cd *hivev1.ClusterDeployment, kubeconfigBytes []byte) error {
+	data := map[string][]byte{
+		"kubeconfig": kubeconfigBytes,
+	}
+	return r.createOrUpdateClusterCredentialSecret(ctx, cd, KubeconfigSecretName(cd.Name), data, kubeconfig)
+}
+
+func (r *Credentials) EnsureAdminPasswordSecret(ctx context.Context, cd *hivev1.ClusterDeployment) (string, error) {
+	secretRef := types.NamespacedName{Namespace: cd.Namespace, Name: KubeadminPasswordSecretName(cd.Name)}
+	secret := &corev1.Secret{}
+	err := r.Get(ctx, secretRef, secret)
+	if err == nil {
+		password, exists := secret.Data["password"]
+		if !exists {
+			r.Log.Warn("failed to find password in secret, generating new one")
+		} else {
+			passwordHash, err := bcrypt.GenerateFromPassword(password, bcrypt.DefaultCost)
+			if err != nil {
+				return "", fmt.Errorf("failed to generate password hash: %w", err)
+			}
+			return string(passwordHash), nil
+		}
+	} else if !errors.IsNotFound(err) {
+		return "", fmt.Errorf("failed to get kubeadmin password secret: %w", err)
+	}
+	r.Log.Infof("Generating admin password")
+	password, err := generateKubeadminPassword()
```

Comment:
Agree, I'll update `EnsureKubeconfigSecret` to work the same.
Mind if it's in a separate PR? it will take some refactoring that will make the PR more complicated than it should be.

Reply:
NM, updated this PR with the change